blogposts/2021/how-we-built-batch-changes.md

```diff
@@ -0,0 +1,213 @@
+"[Aa]utomation" to "[Cc]ampaigns" in our codebase.
+
+What followed were months of fine-tuning, user testing, shipping, tweaking -
+_iterating_. We added more features, fixed a lot of bugs, made things go faster.
+
+But - and by now you should know where this is going - something felt _off_.
+While it worked and we even had customers using it and saying they loved it, we
+also noticed that our colleagues had problems using it, but often being relucant
+to admit that they were confused with the workflow. Some of the questions we
+got: why JSON (yes, we made users write JSON by hand)? Why can't I put
+everything in that JSON file, why do I have to switch to the web UI? And what if
+I want to update the changes?
+
+I wouldn't say we were stuck - we were constantly shipping new features and
+improvements. But we also realised that maybe we've exhausted one idea. Tweaking
+wouldn't get us out of this, nor a sledgehammer. But maybe looking at it from a
+different angle: what if we leave how diffs are produced alone, but instead
+of making users say what should happen to produce a campaign, we switch to a
+_declarative_ model and let them describe what the campaign should like in a
+YAML file?
+
+[This pull request](https://github.com/sourcegraph/sourcegraph/pull/10921) shows
+how we asked "what if?": we wrote the documentation first (yes, README-driven
+development) and showed it to teammates and colleagues. The time it took a
+colleague to go from "ok, tell me what campaigns are" to "ohhh, I get it! nice"
+was reduced tenfold. Very much to our relief, since the changes necessary to
+make the documentation a reality were huge. We had to build a distributed,
+_declarative_ system that manages hundreds and thousands of pull and merge
+requests across different code hosts.
+
+Once the idea was validated by a lot "this totally makes sense" from potential
+users we got to work. Look at [this PR][pr4b], number 4b in this series if you
+will, to see an example of what "declarative" means and how we reconcile the
+current and the desired state (as described by a user) of a changeset. (Sidenote
+for the curious: the code today is [even cooler][evencoolercode], take a look).
+
+That brings us to the last of the five pull requests. 
+
+Towards the end of last year we ripped the "beta" label off of Campaigns and
+started to concentrate on getting more customers to use it: writing better (or
+even any) documentation, improving the onboarding process, providing
+troubleshooting help, fixing bugs and edge cases.
+
+At the same time our fast growing product marketing team started working towards
+making our product as a whole more consistent. One thing that stood out was the
+name "Campaigns" that didn't fit next to the names of our other features and
+always required a sentence or two of explanation ("Glad you asked. A campaign is
+a ..."). The name Batch Changes, on the other hand, fared far better in surveys:
```

Comment:
Maybe this isn't the time or place but could be cool to throw in a few sentences about the naming process. Or at least a couple other options we were considering!

Reply:
"Clusters" and "Bulk Changes" and "Batches" and "Campaigns" were other options and didn't do as well. I'll try to work that in!

A separate "Naming is hard, or: how to find a product name" blog post would be cool.

Reply:
Changed that paragraph